cryptoprotocol/src/main/java/de/upb/crypto/incentive/cryptoprotocol/model/Token.java

```diff
@@ -1,4 +1,135 @@
+ * @author Patrick Schürmann
+ * TODO compute representation of token for serialization
+ */
+public class Token
+{
+    private Zn myRemainderClassRing; // the remainder class ring the token's exponents live in
+    private SPSEQSignatureScheme signatureScheme; // the instance of the signature scheme that is used to certify tokens
+
+    private GroupElement token; // the Pedersen commitment computed from the bases and the exponents, representing the actual token
+    private GroupElementPlainText tokenPlainText; // plaintext version of token, needed for compliance with verification method
+    private SPSEQSignature certificate; // the SPS-EQ certifying the commitment as well-formed and valid
+    private VerificationKey verificationKey; // the SPS-EQ pk used to check validity of token-certificate pair
+
+    private GroupElement[] commitmentBases; // array of the 7 group elements (h_i) needed to form the Pedersen commitment
+    private ZnElement userSecretKey; // secret key of the user the token belongs to
+    private ZnElement encryptionSecretKey; // secret key used for the ElGamal encryption in the Spend algorithm
+    private ZnElement doubleSpendRandomness0; // randomness used for the first challenge generation in double spending protection
+    private ZnElement doubleSpendRandomness1; // randomness used for the second challenge generation in double spending protection
+    private ZnElement points; // number of points that the token currently stores (initially 0)
+    private ZnElement z, t; // values for blinding the token group element
+
+    /**
+     * standard constructor, intializes an empty token, meaning a token storing 0 points represented as a Pedersen commitment.
+     * Note that certificate is not set since a token is usually not certified upon creation.
+     * @param myRemClassRing remainder class ring of exponents
+     * @param sigScheme signature scheme
+     * @param vKey verfication key
+     * @param h array of the bases for the Pedersen commitment
+     * @param USK secret key of user
+     * @param esk ElGamal secret key
+     * @param dsrnd0 double spending randomness
+     * @param dsrnd1 other double spending randomness
+     */
+    public Token(Zn myRemClassRing, SPSEQSignatureScheme sigScheme, VerificationKey vKey,GroupElement[] h, ZnElement USK, ZnElement esk, ZnElement dsrnd0, ZnElement dsrnd1, ZnElement z, ZnElement t) throws PedersenException
+    {
+        // asserting correct number of group elements passed
+        if(h.length != 7)
+        {
+            throw new PedersenException("h is required to consist of 7 group elements, found: " + String.valueOf(h.length));
+        }
+
+        // TODO check out group element vector, vector exponentiation operation
+        // initializing bases array with deep copy of passed elements array
+        this.commitmentBases = new GroupElement[h.length];
+        for(int i = 0; i < h.length; i++)
```

Comment:
I also want to mention the existence of `RingElementVector`, which is basically an immutable array of (in your case) `ZnElement`s. Also nice because it has a bunch of useful methods in conjunction with `GroupElementVector` (e.g., `groupElems.innerProduct(znElems)` computes the standard `\prod g_i^h_i` thing)

Reply:
Okay, I'm changing my answer to "don't pass `h` here at all". Instead, pass something like the Provider's public key if needed.